### ISSUE TITLE
Cross Entropy Loss causes programs to crash in MPS

### ISSUE THREAD

**Issue by cuttle-fish-my** [pytorch/pytorch#116095]
### 🐛 Describe the bug

Hi! I was trying to debug in my m2 macmini and found that the CE loss will make the program crash when the two vectors are in the MPS device. 

Here is a toy example (it works fine in cpu):

```python
import torch
from torch.nn.modules import CrossEntropyLoss

dev = 'mps'

pred = torch.randn(3, 5, requires_grad=True).to(dev)
target = torch.ones(3, dtype=torch.long).to(dev)

loss = CrossEntropyLoss()
output = loss(pred, target)
output.backward()
print(output)
```
and here is the console output:
```
Process finished with exit code 138 (interrupted by signal 10:SIGBUS)
```

However, I found the program works again after converting `target` into one-hot representation:
```python
import torch
from torch.nn.modules import CrossEntropyLoss

dev = 'mps'

pred = torch.randn(3, 5, requires_grad=True).to(dev)
target = torch.ones(3, dtype=torch.long)
target = torch.eye(5)[target].to(dev)

loss = CrossEntropyLoss()
output = loss(pred, target)
output.backward()
print(output)
```
As you can see, it prints the results as normal:
```
tensor(2.4638, device='mps:0', grad_fn=<DivBackward1>)

Process finished with exit code 0
```

Thanks in advance if you can help me!

### Versions

PyTorch version: 2.0.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.2 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.1.0.2.5)
CMake version: version 3.25.2
Libc version: N/A

Python version: 3.11.5 (main, Aug 24 2023, 12:23:19) [Clang 15.0.0 (clang-1500.0.40.1)] (64-bit runtime)
Python platform: macOS-14.2-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2

Versions of relevant libraries:
[pip3] numpy==1.26.2
[pip3] torch==2.0.1
[pip3] torchvision==0.15.2
[conda] Could not collect

cc @ezyang @gchanan @zou3519 @kadeng @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr

**Comment by mikaylagawarecki** [comment]
Marking high priority for crash, could you check whether this issue persists on the latest version of PyTorch (2.1.2) please

**Comment by cuttle-fish-my** [comment]
> Marking high priority for crash, could you check whether this issue persists on the latest version of PyTorch (2.1.2) please

Thanks for replaying! Actually, I've already tried torch>=2.10, and the program works in MPS. But for some reason, I have to use torch==2.0.1 in my project, so it is really torturing...

**Comment by malfet** [comment]
Works for me in 2.1, there is not much we can do to fix 2.0 crash

**Comment by zygi** [comment]
This still crashes on 2.1.1 when using float16:
```py
import torch
from torch.nn.modules import CrossEntropyLoss

dev = 'mps'
dtype=torch.float16

pred = torch.randn(3, 5, requires_grad=True, dtype=dtype).to(dev)
target = torch.ones(3, dtype=torch.long).to(dev)

loss = CrossEntropyLoss()
output = loss(pred, target)
output.backward()
print(output)
```
crashes with
```
(mpsFileLoc): /AppleInternal/Library/BuildRoots/0032d1ee-80fd-11ee-8227-6aecfccc70fe/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm:294:0: error: 'mps.select' op failed to verify that all of {true_value, false_value, result} have same element type
(mpsFileLoc): /AppleInternal/Library/BuildRoots/0032d1ee-80fd-11ee-8227-6aecfccc70fe/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm:294:0: note: see current operation: %10 = "mps.select"(%9, %8, %7) : (tensor<*xi1>, tensor<*xf16>, tensor<1xf32>) -> tensor<*xf16>
/AppleInternal/Library/BuildRoots/0032d1ee-80fd-11ee-8227-6aecfccc70fe/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphComputePackage.mm:180: failed assertion `expected a valid model URL'
[1]    51613 abort      python crash_test.py
```

As above, converting the second argument into a float16 onehot tensor makes the program run without crashing.

**Comment by malfet** [comment]
Ok, this is a bug in MPS framework itself, i.e. [`-divisionNoNaNWithPrimaryTensor:secondaryTensor:name: `](https://developer.apple.com/documentation/metalperformanceshadersgraph/mpsgraph/3675593-divisionnonanwithprimarytensor) generates 
```
    %14 = "mps.constant"() <{value = dense<0.000000e+00> : tensor<1xf32>}> : () -> tensor<1xf32>
    %15 = "mps.divide"(%12, %13) : (tensor<1xf16>, tensor<1xf16>) -> tensor<1xf16>
    %16 = "mps.cast"(%13) <{resultElementType = i1}> : (tensor<1xf16>) -> tensor<1xi1>
    %17 = "mps.select"(%16, %15, %14) : (tensor<1xi1>, tensor<1xf16>, tensor<1xf32>) -> tensor<1xf16>
```
